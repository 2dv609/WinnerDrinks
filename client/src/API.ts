import axios, { AxiosResponse } from "axios"

const baseUrl: string = "http://localhost:4000/"

export const getTrivia = async(): Promise<AxiosResponse<ApiDataType>> => {
  
  try{
    const questions : AxiosResponse<ApiDataType> = await axios.get(baseUrl + 'trivia')

    return questions
  } catch (error)
  {
    throw new Error(error)    
  }    
}
