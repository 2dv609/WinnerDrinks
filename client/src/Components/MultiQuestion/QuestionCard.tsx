import React, { useEffect, useState } from 'react'
import './QuestionCard.css';

type Props = {
  answer: string;
  handleAnswer: any;
}

const QuestionCard: React.FC<Props> = ({ answer, handleAnswer }) => {

  useEffect(() => {
  }, [])

  return (
    <div>
      <div className={'questionDiv'}
        onClick={handleAnswer}>
        {answer}
      </div>
    </div>
  );
}

export default QuestionCard;