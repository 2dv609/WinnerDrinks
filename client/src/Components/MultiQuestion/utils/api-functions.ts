import fetch from 'node-fetch';

export const formatAPIResponseString = (input: string) => {
  let spanTag = document.createElement('span')
  spanTag.innerHTML = input
  return spanTag.innerHTML
}

export const fetchQuestions = async (questionAmount: number) => {
  const fetchResponse = await fetch(`https://opentdb.com/api.php?amount=${questionAmount}`);
  const jsonFetchResponse = await fetchResponse.json();
  return jsonFetchResponse.results;
}