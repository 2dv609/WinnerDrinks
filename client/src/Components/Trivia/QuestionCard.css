/* .questionDiv {
  border: 2px solid;
  border-radius: 5px;
  width: 250px;
  text-align: center;
  margin-right: 5px;
  background-color: #91ccff;
}
 */
