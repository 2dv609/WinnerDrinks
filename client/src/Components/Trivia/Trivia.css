/* .answersDiv {
  display: flex;
  justify-content: center;
  flex-flow: wrap;
  max-width: 600px;
  margin: 0 auto;
}

.playerCardInactive {
  background-color: rgb(97, 0, 0);
}

.playerCardActive {
  background-color: green;
} */