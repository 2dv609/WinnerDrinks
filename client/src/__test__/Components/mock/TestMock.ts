import { GameModuleProps } from '../../../Components/GameModuleProps'
import Player from '../../../model/Player'

export const playersMock: Player[] = [new Player('Tom'), new Player('Carl'), new Player('Sandra')]

const makeWinnerAlertMock = (p: Player | Player[] | null, message?: string): void => {}
const chooseRandomNewGameMock = (): void => {}
const removeGameEventMock = (gameEventId: string, gameEvents: GameEventAPI): void => {}
const getPlayersMock = (amount: number, players: Player[]): Player[] => {return players}
const addScoreMock = (p: Player) => void {}


export const gameServiceMock: GameModuleProps = {
    players: playersMock,
    getPlayers: getPlayersMock,
    addScore: addScoreMock,
    makeWinnerAlert: makeWinnerAlertMock,
    chooseRandomNewGame: chooseRandomNewGameMock,
    removeGameEvent: removeGameEventMock        
}
