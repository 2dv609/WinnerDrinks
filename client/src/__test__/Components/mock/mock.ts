import Player from '../../../Player'


// Mock GameProps
const addScoreMock = (p: Player, score: number) => {}
const makeWinnerAlertMock = (p: any) => {}
const chooseRandomNewGameMock = () => {}
const getPlayersMock = (amount: number): Player[] => {
    const player1 = new Player('Mats')
    const player2 = new Player('Tom')
    return [player1, player2];
};

export const gamePropsMock = {
    getPlayers: getPlayersMock, // Let the component decide how many players it needs. 
    addScore: addScoreMock, // Adds a score to a player
    makeWinnerAlert: makeWinnerAlertMock, // makes an Alert that corresponds to if there are one, multiple, or no winners
    chooseRandomNewGame: chooseRandomNewGameMock, // chooses a random new game
}

// Mock game event mock
export const gameEventMock = {
    _id: 111,
    question: 'This is a mock question'
}