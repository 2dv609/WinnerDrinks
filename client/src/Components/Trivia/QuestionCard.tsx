import React, { useEffect } from 'react'
import './QuestionCard.css';

type Props = {
  answer: string;
  handleAnswer: (e: any) => void;
}

const QuestionCard: React.FC<Props> = ({ answer, handleAnswer }) => {

  useEffect(() => {
  }, [])

  return (

      <div className={'questionDiv card'}
        onClick={handleAnswer}>
        {answer}
      </div>

  );
}

export default QuestionCard;