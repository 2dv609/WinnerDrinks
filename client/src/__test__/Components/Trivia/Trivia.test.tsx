import React from 'react';
import ReactDOM from 'react-dom';
import Trivia from '../../../Components/Trivia/Trivia'
import QuestionCard from '../../../Components/Trivia/QuestionCard'
import { render, fireEvent } from '@testing-library/react' 
import '@testing-library/jest-dom/extend-expect'
import { getGameService } from '../../../model/ModelFactory'
import GameService from '../../../model/GameService'
import Player from '../../../model/Player'
import API from '../../../util/API'
import { playersMock, gameServiceMock } from '../mock/TestMock'
import { formatAPIResponseString } from '../../../Components/Trivia/utils/api-functions'
import { v1 as uuidv1 } from 'uuid'

/* --------------------------------- */
/* Test cases for game module trivia. */
/* --------------------------------- */

describe('Test suite for game module party', () => {

    /* --------------------------------- */
    /* ----- Test cases setup ---------- */
    /* --------------------------------- */

    const api: API = new API()
    let triviaEvents: GameEventAPI | undefined
    
    const gameService: GameService = getGameService()
    const currentPlayer: Player[] = gameService.getPlayers(1, playersMock)
    const handleAnswerMock = (e: any) => {
        gameServiceMock.addScore(currentPlayer[0])
        gameServiceMock.makeWinnerAlert(null)
        gameServiceMock.chooseRandomNewGame()
    }
    const triviaQuestionMock: IParty | ITrivia | IBackToBack = {
        _id: 'mockID',
        category: 'mock',
        type: 'multiple',
        difficulty: 'easy',
        question: 'What is not a wind instrument?',
        correct_answer: 'Viola',
        incorrect_answers: [
          'Oboe',
          'Trombone',
          'Duduk'
        ]
    }

    beforeAll(async () => {
        triviaEvents = await api.getGameEvents('trivia')
    })

    test('Game module trivia should use props gameService, gameEvent and currentPlayers', () => {
        const div = document.createElement('div')
        ReactDOM.render(<Trivia currentPlayers={currentPlayer} gameService={gameServiceMock} gameEvent={triviaQuestionMock}></Trivia>, div)
    })

    test('T1.MT.S.1: Trivia game events should contain at least 20 game events', () => {
        if (!triviaEvents) {
            return
        }

        expect(triviaEvents.questions.length >= 20).toBeTruthy()
    })

    test('T1.MT.S.2: Trivia game events should contain only unique game events', () => {
        if (!triviaEvents) {
            return
        }

        const questions: string[] = []
        
        triviaEvents.questions.forEach((event: IParty | ITrivia | IBackToBack ) => {
            questions.push(event.question)
        })

        expect(questions.length === new Set(questions).size).toBeTruthy()
    })

    test('T1.MT.S.3: Trivia game events should not have more than 60 words', () => {
        if (!triviaEvents) {
            return
        }

        triviaEvents.questions.forEach((event: IParty | ITrivia | IBackToBack ) => {
            expect(event.question.split(' ').length <= 60).toBeTruthy()
        })
    })


    test('T1.MT.UI.1: Trivia game events should contain 1 active game participant\'s name', () => {
        const { getByTestId } = render(<Trivia currentPlayers={currentPlayer} gameService={gameServiceMock} gameEvent={triviaQuestionMock}></Trivia>)
        expect(getByTestId('current-player')).toHaveTextContent(currentPlayer[0].name)
        
    })

    test('T1.MT.UI.2: Game module trivia should display a trivia game event', () => {
        const { getByTestId } = render(<Trivia currentPlayers={currentPlayer} gameService={gameServiceMock} gameEvent={triviaQuestionMock}></Trivia>)
        expect(getByTestId('current-player')).toHaveTextContent(currentPlayer[0].name)
        expect(getByTestId('game-event')).toHaveTextContent(triviaQuestionMock.question)

    })

    test('T1.MT.UI.3: Select a trivia event winner should trigger transition to next game round, button 0', () => {
        const { getByTestId } = render(<QuestionCard answer={formatAPIResponseString('mockAnswer')} handleAnswer={handleAnswerMock} key={uuidv1()}></QuestionCard>)
        
        const addScoreSpy = jest.spyOn(gameServiceMock, 'addScore')
        const makeWinnerAlertSpy = jest.spyOn(gameServiceMock, 'makeWinnerAlert')
        const chooseNewRandomGameSpy = jest.spyOn(gameServiceMock, 'chooseRandomNewGame')
        
        fireEvent.click(getByTestId('answer-alternatives'))
        expect(addScoreSpy).toBeCalled()
        expect(makeWinnerAlertSpy).toBeCalled()
        expect(chooseNewRandomGameSpy).toBeCalled()
    })
})
