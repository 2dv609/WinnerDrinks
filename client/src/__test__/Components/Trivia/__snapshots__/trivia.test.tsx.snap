// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game module should match snapshot 1`] = `
<div
  className="box"
>
  <div
    id="gameInfo"
  >
    <h4>
      Turn to answer a question: 
    </h4>
    <ul>
      <li
        className="card"
      >
        Mats
      </li>
    </ul>
  </div>
  <br />
  <h4>
    Loading questions
  </h4>
</div>
`;
